dataland-e2etests/src/test/kotlin/org/dataland/e2etests/utils/ApiAccessor.kt

```diff
@@ -122,7 +124,32 @@ class ApiAccessor {
                 )
             }
         }
-        return listOfUploadInfo
+        if (!ensureQaPassed) return listOfUploadInfo
+        return ensureQaCompleted(listOfUploadInfo)
+    }
+
+    private fun ensureQaCompleted(listOfUploadInfo: MutableList<UploadInfo>): MutableList<UploadInfo> {
+        // Check after upload to ensure that dummy QA is passed and wait until it is
+        val maxQaPassedYetRetries = 100; val sleepIfQaNotPassedYetMs: Long = 50
+
+        val result = mutableListOf<UploadInfo>()
+        repeat(maxQaPassedYetRetries) {
+            for (idx in listOfUploadInfo.indices.reversed()) {
+                var metaData = listOfUploadInfo[idx].actualStoredDataMetaInfo!!
+                listOfUploadInfo[idx].actualStoredDataMetaInfo = metaDataControllerApi.getDataMetaInfo(metaData.dataId)
+                metaData = listOfUploadInfo[idx].actualStoredDataMetaInfo!!
+                if (metaData.qaStatus == QAStatus.accepted) {
+                    result.add(listOfUploadInfo[idx])
+                    listOfUploadInfo.removeAt(idx)
+                }
+            }
+            if (listOfUploadInfo.isEmpty()) return result
+            Thread.sleep(sleepIfQaNotPassedYetMs)
+        }
+        class QaNotCompletedException(message: String) : RuntimeException(message)
+        throw QaNotCompletedException(
+            "$maxQaPassedYetRetries retries every $sleepIfQaNotPassedYetMs ms, qa for $listOfUploadInfo still failed",
+        )
```

Comment:
What about this approach: 
A list of all dataIds is collected. Now it's checked for every ID if QA is passed. If not, this ID is put to a new list. 
In the End, the new list is assigned to the old list and list.length() == 0 is returned. 

this method now goes into Awaitility

Reply:
Was refactored in the meantime due to Sonarcube complains. 

Operating on list of dataIds is difficult since UploadInfo has a lot more than just the dataIds and keeping track between the mapping of each dataId to the associated UploadInfo would require quite some effort. Therefore I've instead changed the metaData in UploadInfo to be a var and operate on UploadInfo objects directly.

Other than this the done refactoring into method updateQaStatus seems similar to what you propose.

Reply:
Has been refactored as proposed

Reply:
ok